e2etest/clean/clean_ec2.go

```diff
@@ -0,0 +1,70 @@
+package main
+
+import (
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"log"
+	"time"
+)
+
+func main() {
+	// set up aws go sdk ec2 client
+	testSession, err := session.NewSession()
+	if err != nil {
+		log.Fatalf("Error creating session %v", err)
+	}
+	ec2client := ec2.New(testSession)
+
+	// Get list of instance
+	//State filter
+	instanceStateName := "instance-state-name"
+	instanceStateValue := "running"
+	instanceStateValues := []*string{&instanceStateValue}
+	instanceStateFilter := ec2.Filter{Name: &instanceStateName, Values: instanceStateValues}
+
+	//Tag filter
+	instanceTagName := "tag:Name"
+	instanceTagValueIntegTestSampleApp := "Integ-test-Sample-App"
+	instanceTagValueIntegTestAOC := "Integ-test-aoc"
+	instanceTagValues := []*string{&instanceTagValueIntegTestSampleApp, &instanceTagValueIntegTestAOC}
+	instanceTagFilter := ec2.Filter{Name: &instanceTagName, Values: instanceTagValues}
+
+	//get instances to delete
+	var deleteInstances []ec2.Instance
+	var nextToken *string
+	for {
+		describeInstancesInput := ec2.DescribeInstancesInput{Filters: []*ec2.Filter{&instanceStateFilter, &instanceTagFilter}, NextToken: nextToken}
+		describeInstancesOutput, err := ec2client.DescribeInstances(&describeInstancesInput)
+		if err != nil {
+			log.Fatalf("Failed to get instance for error %v", err)
+		}
+		for _, reservation := range describeInstancesOutput.Reservations {
+			for _, instance := range reservation.Instances {
+				//only delete instance if older than 30 days
+				if time.Now().UTC().Add(-1 * time.Hour * 24 * 30).After(*instance.LaunchTime) {
+					deleteInstances = append(deleteInstances, *instance)
+				}
+			}
+		}
+		if describeInstancesOutput.NextToken == nil {
+			break
+		}
+		nextToken = describeInstancesOutput.NextToken
+	}
+
+	//delete instances
+	var instanceIds []*string
+	for _, instance := range deleteInstances {
+		instanceIds = append(instanceIds, instance.InstanceId)
+		log.Printf("Delete instance %s tags %s launch-date %s", *instance.InstanceId, instance.Tags, instance.LaunchTime)
+	}
+	if len(instanceIds) < 1 {
```

Comment:
Isn't this directly proportional to the length of `deleteInstances`? We can swap these around and put the if check before the for loop to short circuit earlier, though that's just a nitpick.